crates/env_filter/src/parser.rs

```diff
@@ -42,20 +77,26 @@ pub(crate) fn parse_spec(spec: &str) -> (Vec<Directive>, Option<FilterOp>) {
                         if let Ok(num) = part1.parse() {
                             (num, Some(part0))
                         } else {
-                            eprintln!(
-                                "warning: invalid logging spec '{}', \
+                            err_or_print(
+                                skip_errors,
+                                format!(
+                                    "warning: invalid logging spec '{}', \
                                  ignoring it",
-                                part1
-                            );
+                                    part1
+                                ),
+                            )?;
                             continue;
                         }
                     }
                     _ => {
-                        eprintln!(
-                            "warning: invalid logging spec '{}', \
+                        err_or_print(
```

Comment:
Not too thrilled with `err_or_print`

I'm hoping we can instead have the following commits
- refactor commit to change `parse_spec` to return an error with `Builder::parse` taking the error and printing it as a warning
- feature commit that pulls the code out of `Builder::parse` into `Builder::try_parse`, making `Builder::parse` a wrapper around it that takes the error and prints it as a warning

Reply:
I agree, this was not the best solution. But `parse_spec` cannot just return `Result`, as `Builder::parse` method needs the valid part of the spec even in case of an error. Also, it is possible to have multiple errors in the same spec, so just bailing out with the first error would also not work.

Instead, I've added a new `ParseResult` struct, that contains original returned values plus the encountered errors, and made the caller decide, what they want to do with these errors. I believe this approach looks quite clean now.